File: no_1231/sol.py
class Solution:
    def maximizeSweetness(self, sweetness: List[int], k: int) -> int:
        low = 1 #초기화 
        print("sum(sweetness): ", sum(sweetness))
        print("k + 1: ", k + 1)
        high = int(sum(sweetness) / (k+1)) # 초기화
        print("low : ", low)
        print("high: ", high)
        while low < high:
            mid = math.floor((high + low + 1) / 2)
            if self.can_split(sweetness, k, mid):
                low = mid
            else:
                high = mid - 1
        return low
        
    def can_split(self, sweetness: List[int], k: int, mid: int) -> bool:
        temp_sum = 0
        chunks = 0
        for element in sweetness:
            temp_sum += element
            if temp_sum >= mid:
                chunks += 1
                temp_sum = 0
        return chunks >= k + 1      

                
                

File: no_1231/sol.java
class Solution {
    public int maximizeSweetness(int[] sweetness, int k) {
        int low = 1, high = Arrays.stream(sweetness).sum() / (k + 1);
        System.out.println("high: "+high);
        while (low < high){
            int mid = (low + high + 1) / 2;
            if (canSplit(sweetness, k, mid)) low = mid;
            else high = mid - 1;
        }
        return low;     
    }
                
    private boolean canSplit(int[] sweetness, int k, int mid) {
        int chunks = 0, sum = 0;
        for (int i = 0; i < sweetness.length; i++){
            sum += sweetness[i];
            if (sum >= mid){
                ++chunks;
                sum = 0;
            }
        }
        return chunks >= k + 1;
    }
}
